File: backend/voting-app/src/main/java/com/tu/votingapp/utils/mappers/referendum/ReferendumMapper.java
package com.tu.votingapp.utils.mappers.referendum;

import com.tu.votingapp.dto.general.referendum.ReferendumDTO;
import com.tu.votingapp.entities.referendum.ReferendumEntity;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;

@Mapper(componentModel = "spring", uses = {ReferendumOptionMapper.class})
public interface ReferendumMapper {

    // Map entity to DTO: extract createdBy.id into createdById.
    @Mapping(source = "createdBy.id", target = "createdById")
    ReferendumDTO toDto(ReferendumEntity entity);

    // Map DTO to entity: create a minimal UserEntity using the createdById.
    @Mapping(target = "createdBy", // Target field only
            // Use expression to create the UserEntity, accessing createdById from the dto
            // Ensure UserEntity has a public UserEntity(Long id) constructor
            // Added null check for robustness
            expression = "java(dto.getCreatedById() == null ? null : new com.tu.votingapp.entities.UserEntity(dto.getCreatedById()))")
    ReferendumEntity toEntity(ReferendumDTO dto);
}

File: frontend/android/app/src/main/java/com/example/android/utils/mappers/ReferendumMapper.kt
package com.example.android.utils.mappers

import com.example.android.dto.response.referendum.OptionResponseDTO
import com.example.android.dto.response.referendum.ReferendumResponseDTO
import com.example.android.entity.referendum.ReferendumEntity

object ReferendumMapper {
    fun ReferendumResponseDTO.toEntity(): ReferendumEntity = ReferendumEntity(
        id = this.id,
        title = this.title,
        description = this.description,
        question = this.question,
        startDate = this.startDate,
        endDate = this.endDate,
        status = this.status,
        createdById = this.createdById
    )

    fun ReferendumEntity.toDto(options: List<OptionResponseDTO>): ReferendumResponseDTO = ReferendumResponseDTO(
        id = this.id!!,
        title = this.title,
        description = this.description,
        question = this.question,
        startDate = this.startDate,
        endDate = this.endDate,
        status = this.status,
        createdById = this.createdById,
        options = options
    )
}